### PR TITLE
GTJ-15

### PR DESCRIPTION
**Pull request by kkeszthe**
Sorry, for all the files

### CODE REVIEW

File: backend/src/controllers/troopsController.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { TroopsService } from '../services';
+import { getTroopsForKingdom, insertTroopForKingdom } from '../repos/troops';
```
Comment (adamgyulavari) [inline backend/src/controllers/troopsController.js:2]: @kkeszthe these should be in the `TroopsRepo` class too, and either import the whole class and make an instance or export the instance from there and import that here

File: backend/src/repos/troops.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  defence,
+  started_at,
+  finished_at
+) => {
+  const troop = await db.query(
+    `update kingdom_troops set level=?, hp=?, attack=?, defence=?, started_at=?, finished_at=? where kingdom_id=? LIMIT 1;`,
+    [level, hp, attack, defence, started_at, finished_at, kingdomID]
+  );
+  return troop.results;
+};
```
Comment (adamgyulavari) [inline backend/src/repos/troops.js:41]: @kkeszthe so this should be a class too

File: backend/src/services/TroopsService.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+const setMinutes = async (startTime, minutes) => {
+  return startTime.setUTCMinutes(startTime.getUTCMinutes() + minutes);
+};
```
Comment (adamgyulavari) [inline backend/src/services/TroopsService.js:3]: @kkeszthe why is this here?

File: backend/src/services/TroopsService.test.js [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+      },
+      {
+        id: 2,
+        level: 1,
+        hp: 1,
+        attack: 1,
+        defence: 1,
+        started_at: '2020-07-04T09:45:00.000Z',
+        finished_at: '2020-07-04T09:46:00.000Z',
+      },
```
Comment (adamgyulavari) [inline backend/src/services/TroopsService.test.js:40]: @kkeszthe these things are making the tests very long, let's move them into factories:
`TroopFactory.create({id: 1})` creates the first troop,
similarly with resources `ResourceFactory.create({type: 'gold', amount: 200})`

File: backend/src/services/buildingService.js [hunk 5]
```diff
@@ -6,86 +6,96 @@ const errorMessages = {
...
+      await resourceService.spendGold(input.kingdomId, resourcesData.cost);
+      await resourceService.updateGoldGeneration(
+        input.kingdomId,
+        resourcesData.genAdd
+      );
+      break;
+    default:
+      break;
+  }
+  return formatData(buildingId, buildingDataInput);
```
Comment (adamgyulavari) [inline backend/src/services/buildingService.js:95]: @kkeszthe these changes will spark real joy in @gulyaasjaanos eyes, because they will cause like the whole file to be conflicting :D

File: backend/src/services/userService.js [hunk 6]
```diff
@@ -1,56 +1,64 @@
...
+  } else {
+    throw { error: 'Username or password is incorrect.' };
+  }
+};
+
+export const userService = {
+  filterInput,
+  add,
+  get,
+};
```
Comment (adamgyulavari) [inline backend/src/services/userService.js:64]: @kkeszthe and these too

**Review by adamgyulavari — CHANGES_REQUESTED** [review]

File: backend/src/services/buildingService.js (hunk 5 above)
Comment (kkeszthe) [reply backend/src/services/buildingService.js:95]: @adamgyulavari do you expect action from me here or is it just a warning that I should not commit such changes in the future?

Comment (adamgyulavari) [reply backend/src/services/buildingService.js:95]: the latter :D

File: backend/src/repos/TroopsRepo.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import { db } from '../data/connection';
+
+class TroopsRepo {
+  async get(kingdomID) {
+    console.log('repo');
+    console.log(kingdomID);
```
Comment (adamgyulavari) [inline backend/src/repos/TroopsRepo.js:6]: @kkeszthe remove logs

File: backend/src/repos/TroopsRepo.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    const troops = await db.query(
+      `select * from kingdom_troops where kingdom_id = ?;`,
+      kingdomID
+    );
+    return troops.results;
+  }
+
+  async insert(kingdomID, level, hp, attack, defence, started_at, finished_at) {
+    const troop = await db.query(
+      `insert into kingdom_troops (kingdom_id, level, hp, attack, defence, started_at, finished_at) values(?,?,?,?,?,?,?);`,
```
Comment (adamgyulavari) [inline backend/src/repos/TroopsRepo.js:16]: @kkeszthe we should extract the sql table field names into constants, so if the name is changed, we need to rewrite only once

**Review by adamgyulavari — CHANGES_REQUESTED** [review]

File: backend/src/repos/TroopsRepo.js [hunk 9]
```diff
@@ -1,28 +1,35 @@
...
     return troop.results;
   }
 
-  async update(kingdomID, level, hp, attack, defence, started_at, finished_at) {
+  async update(kingdomID, level, started_at, finished_at) {
+    const columns = await this.getColumnNames('kingdom_troops');
     const troop = await db.query(
-      `update kingdom_troops set level=?, hp=?, attack=?, defence=?, started_at=?, finished_at=? where kingdom_id=? LIMIT 1;`,
-      [level, hp, attack, defence, started_at, finished_at, kingdomID]
+      `update kingdom_troops set ${columns[2]} = ${columns[2]} + 1, ${columns[3]} = ${columns[3]} + 1, ${columns[4]} = ${columns[4]} + 1, ${columns[5]} = ?, ${columns[6]}=?, ${columns[7]}=? where ${columns[1]} = ? and ${columns[2]} = ? LIMIT 1;`,
```
Comment (adamgyulavari) [inline backend/src/repos/TroopsRepo.js:31]: @kkeszthe this looks weird `${columns[2]} = ${columns[2]} + 1` are you sure?

File: backend/src/repos/getColumnNames.js [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+export const getColumnNames = async table => {
+  const columns = await (await db.query(`show columns from ${table};`)).results;
+
+  await columns.forEach((column, index) => {
+    columns[index] = column.Field;
+  });
+
+  return columns;
+};
```
Comment (adamgyulavari) [inline backend/src/repos/getColumnNames.js:11]: @kkeszthe oooh, I see what you did there :D
no, this can be in the code, you'll have to modify the repo significantly if there are changes in the db because of this, with my suggestion it'll work, just maybe it won't use the new columns

**Review by adamgyulavari — CHANGES_REQUESTED** [review]

File: backend/src/repos/TroopsRepo.js [hunk 11]
```diff
@@ -1,28 +1,35 @@
 import { db } from '../data/connection';
+import { getColumnNames } from './getColumnNames';
```
[the remark this answers is not in the document]
Comment (kkeszthe) [reply backend/src/repos/TroopsRepo.js:2]: @adamgyulavari I thought the function is reusable. This function can get column names from any table.

File: backend/src/repos/TroopsRepo.js [hunk 12]
```diff
@@ -1,28 +1,35 @@
...
 
 class TroopsRepo {
+  constructor() {
+    this.getColumnNames = getColumnNames;
+  }
   async get(kingdomID) {
-    console.log('repo');
-    console.log(kingdomID);
     const troops = await db.query(
       `select * from kingdom_troops where kingdom_id = ?;`,
```
[the remark this answers is not in the document]
Comment (kkeszthe) [reply backend/src/repos/TroopsRepo.js]: @adamgyulavari interesting, I could not find this in my previous commit or in my local version either but I hope now it disappears forever.

File: backend/src/repos/TroopsRepo.js [hunk 13]
```diff
@@ -1,34 +1,47 @@
...
       [kingdomID, level, hp, attack, defence, started_at, finished_at]
     );
     return troop.results;
   }
 
   async update(kingdomID, level, started_at, finished_at) {
-    const columns = await this.getColumnNames('kingdom_troops');
     const troop = await db.query(
-      `update kingdom_troops set ${columns[2]} = ${columns[2]} + 1, ${columns[3]} = ${columns[3]} + 1, ${columns[4]} = ${columns[4]} + 1, ${columns[5]} = ?, ${columns[6]}=?, ${columns[7]}=? where ${columns[1]} = ? and ${columns[2]} = ? LIMIT 1;`,
+      `update ${this.tableName}  set ${this.columns.level} = ${this.columns.level} + 1, ${this.columns.hp} = ${this.columns.hp} + 1, ${this.columns.attack} = ${this.columns.attack} + 1, ${this.columns.defence} = ${this.columns.defence} + 1, ${this.columns.started_at}=?, ${this.columns.finished_at}=? where ${this.columns.kingdom_id} = ? and ${this.columns.level} = ? LIMIT 1;`,
```
Comment (adamgyulavari) [inline backend/src/repos/TroopsRepo.js:44]: @kkeszthe final one I hope :D
but in the repo we shouldn't have business logic, that should be in services.
having update here should update the troop anyhow based on a specific troop id with all it's columns
now we have an update that mixes the troop upgrade logic

**Review by adamgyulavari — APPROVED** [review]
